### ISSUE TITLE
The tf.keras.Model.compile metrics do not respect masking since TF 2.2

### ISSUE THREAD

**Issue by foxik** [tensorflow/tensorflow#40002]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): **Yes**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): **Linux Debian Stable**
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: --
- TensorFlow installed from (source or binary): **binary**
- TensorFlow version (use command below): **TF 2.2, TF-nightly 2.3.0-dev20200529**
- Python version: **3.7**

**Describe the current behavior**
The metrics passed to `tf.keras.Model.compile` as `metrics` do not respect model masks.

**Describe the expected behavior**
Until TF 2.1, these metrics did respect model masks.

**Standalone code to reproduce the issue**
Consider the following code which masks the input element.
```python
import numpy as np
import tensorflow as tf
print(tf.__version__)

model = tf.keras.Sequential([
    tf.keras.layers.Masking(1.),
    tf.keras.layers.Dense(1)
])
model.compile(optimizer=tf.optimizers.Adam(),
              loss=tf.losses.MeanSquaredError(),
              metrics=[tf.metrics.MeanSquaredError()],
              weighted_metrics=[tf.metrics.MeanSquaredError()])

print(model.train_on_batch(np.ones([1, 1]), np.ones([1, 1])))
```

TensorFlow until 2.1 masks also the metric in `metrics`, while TensorFlow 2.2 and later do not.
- TensorFlow 2.1 [(colab)](https://colab.research.google.com/drive/1q_CgkAh-wMdXy93ENXUBqU6JKWSPraBi?usp=sharing) prints `[0.0, 0.0, 0.0]`
- TensorFlow 2.2 [(colab)](https://colab.research.google.com/drive/1_iEw5yPbJInfjWAjmgYZijUjvqrzpjTZ?usp=sharing) prints `[0.0, 1.0, 0.0]`
- TensorFlow-Nightly 2.3.0-dev20200529 [(colab)](https://colab.research.google.com/drive/1d4_B_rxfRcPdJcCAFPNw2fV1_VG2kcVv?usp=sharing) prints `[0.0, 1.0, 0.0]`

**Other info / logs**
The logic of applying the mask in `master` is here:
https://github.com/tensorflow/tensorflow/blob/a1ae008076e14f7e445abf2605759779d2a1fb8b/tensorflow/python/keras/engine/compile_utils.py#L404-L414
The `metrics` do not get called with `sample_weight`, but that is the place where the masks are applied (in `apply_mask`).

On the other hand, in TF 2.1
https://github.com/tensorflow/tensorflow/blob/3ffdb91f122f556a74a6e1efd2469bfe1063cb5c/tensorflow/python/keras/engine/training.py#L2000-L2012
the `output_mask` was passed even for the unweighted metrics.

**Comment by Saduf2019** [comment]
I am able to replicate this issue, [masking occurring in 2.1](https://colab.sandbox.google.com/gist/Saduf2019/16ac32138d26f2a15b72c284d30b2442/untitled205.ipynb) and works as not in [tf 2.2](https://colab.sandbox.google.com/gist/Saduf2019/32c69933c866dce75e755d620ad8b663/untitled208.ipynb)

**Comment by goldiegadde** [comment]
@foxik this issue has been fixed in [f66d36cc](https://github.com/tensorflow/tensorflow/commit/1c144fdb678efa94511685872ba7dcccf66d36cc). 
Please reopen if thats not the case.
